### ISSUE TITLE
Surprising 'Type alias "..." cannot refer to itself indirectly in its definition' error with dataclasses

### ISSUE THREAD

**Issue by comex** [microsoft/pyright#4596]
**Describe the bug**
In the example below, the only self-reference comes from a constructor argument type, so it doesn't seem like this should be an error.
This only happens when the class uses `@dataclass`, and does not happen if `init=False` is passed to `dataclass`.  However, `init` should not make a difference since the class defines its own `__init__`.  To quote the Python documentation, "If the class already defines `__init__()`, this parameter [i.e. the `init` parameter to `dataclass`] is ignored."

**To Reproduce**
```
$ pyright /tmp/example.py
[..]
  /tmp/example.py:4:1 - error: Type alias "FooOrInt" cannot refer to itself indirectly in its definition (reportGeneralTypeIssues)
  /tmp/example.py:8:24 - error: Type of parameter "x" is unknown (reportUnknownParameterType)
```
**Expected behavior**
No errors.

**Screenshots or Code**
```python
# example.py
from typing_extensions import TypeAlias
from dataclasses import dataclass

FooOrInt: TypeAlias = 'Foo | int'

@dataclass
class Foo:
    def __init__(self, x: FooOrInt) -> None:
        raise NotImplementedError
```

**VS Code extension or command-line**
pyright 1.1.293, CLI

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.294, which I just published. It will also be included in a future release of pylance.
